src/acl_hostch.cpp

```diff
@@ -682,6 +685,399 @@ clUnmapHostPipeIntelFPGA(cl_mem pipe, void *mapped_ptr, size_t size_to_unmap,
   return CL_SUCCESS;
 }
 
+void acl_read_program_hostpipe(void *user_data, acl_device_op_t *op) {
+
+  cl_event event = op->info.event;
+  cl_int status = 0;
+  size_t pulled_data = 0;
+  bool blocking = event->cmd.info.host_pipe_info.blocking;
+  acl_assert_locked();
+
+  if (!acl_event_is_valid(event) ||
+      !acl_command_queue_is_valid(event->command_queue)) {
+    acl_set_device_op_execution_status(op, -1);
+    return;
+  }
+
+  acl_device_program_info_t *dev_prog =
+      event->command_queue->device->loaded_bin->get_dev_prog();
+  auto host_pipe_info = dev_prog->program_hostpipe_map.at(
+      std::string(event->cmd.info.host_pipe_info.logical_name));
+  acl_mutex_lock(&(host_pipe_info.m_lock));
+  acl_set_device_op_execution_status(op, CL_SUBMITTED);
+  acl_set_device_op_execution_status(op, CL_RUNNING);
+
+  if (host_pipe_info.implement_in_csr) {
+    // CSR read, currently only blocking version is implemented
+    unsigned long long parsed;
+    uintptr_t data_reg, ready_reg, valid_reg;
+    // Convert the CSR address to a pointer
+    try {
+      parsed = std::stoull(host_pipe_info.csr_address, nullptr);
+    } catch (const std::exception &) {
+
+      acl_set_device_op_execution_status(op, -1);
+      return;
+    }
+
+    data_reg = static_cast<uintptr_t>(parsed);
+    ready_reg = static_cast<uintptr_t>(
+        parsed + 8); // ready reg is data reg shift by 8 byte, move this to the
+                     // autodiscovery string maybe
+    valid_reg = static_cast<uintptr_t>(
+        parsed + 16); // valid reg is ready reg shift by 8 byte
```

Comment:
Perhaps instead of 8 and 16 we can set this as preprocessor variable or `static constexpr unsigned`? Though It's not that hard to grasp from the code that this these numbers are offsets so probably also okay to leave as it.